design-proposals/garbage-collect-completed-dvs.md

```diff
@@ -0,0 +1,70 @@
+# Overview
+DataVolumes are a useful construct for provisioning VM disks but once the PVC is provisioned and populated the DataVolume provides no value. This design presents an approach for garbage collecting completed DataVolumes. 
+
+## Motivation
+In the field, customers are confused by DVs after the fact. For example, when someone deletes a PVC that is owned by a DV they are confused when the PVC is re-created.
+
+## Goals
+DataVolumes will be garbage collected after their PVC is provisioned and populated.
+
+## Non Goals
+(limitations to the scope of the design)
+
+## Definition of Users
+(who is this feature set intended for)
+
+## User Stories
+* As a VM owner I use a standalone DataVolume to prepare a VM disk PVC for use by my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. The PVC should still maintain an independent lifecycle from the VM.
+* As a VM owner I add an entry to the dataVolumeTemplates section of my VM spec to have a disk prepared for my VM. Once the DataVolume is complete, it has no use and can be confusing, so I want the system to garbage collect it. I want kubevirt to understand that it doesn't need to create a new DataVolume when the underlying PVC exists. The PVC should still share the same lifecycle as the VM.
+* As a VM owner I want to restore my backed up VM without needing to recreate the DataVolume.
+* As a VM owner I want to replicate my workload to another cluster without needing to mutate the PVC and DV with special annotations in order for them to behave as expected in the new cluster.
+
+## Repos
+* **CDI**: Controllers (DataVolume/Import/Upload/Clone/DataImportCron/DataSource) will need adaptations:
+** So far the controllers assumed the existance of DV-PVC pairs, which is no longer a valid assumption.
+** DataImportCron controller currently re-creates the import DV if deleted, even if the PVC was already populated.
+** Warm imports - verify DV is Succeded only after the last cycle
+* **KubeVirt**: VM controller adaptations (see below)
+* **Tekton tasks** - check how DVs are used to make sure GC by default wouldn't break them
+* **UI**: check if DVs are referred after Succeded
+* **HCO**: expose CDIConfig `dataVolumeTTL`
+
+## Configuration
+* CDIConfig - minimal time for DV garbage collection after completion (disabled by default, for opt-in)
+  `dataVolumeTTL *int32`
+* DataVolume annotation allowing its garbage collection, `true` by default for newly created DVs when `dataVolumeTTL` is set. User can annotate older DVs to get them garbage collected too.
+  `"cdi.kubevirt.io/storage.dv.deleteAfterCompletion": "true"`
```

Comment:
I've shared my opinion here in some of our chats, but i'll refine my thoughts here for a more finalized discussion. 

Changing the default behavior (even only when the feature gate is enabled) introduces risk for users.

For example, if a user has a standalone DV workflow today which expects DVs to exist after PVC population, then we'd break that assumption after an update when the user's workflow creates new DVs which magically disappear on them.

This is the same reason why our virt-controller component in KubeVirt has to be altered to understand this GC feature. We happen to control that KubeVirt controller within this ecosystem, so we know how to make that change, but it's feasible that users outside of our ecosystem have built their own logic around DVs that we are unaware of.

Now all that said. I have no data indicating this user flow I'm concerned about exists or not, however this type of scenario is why we try our best to keep APIs behaving consistently across releases. We're never going to know exactly how people are using our APIs.

There's a risk/reward proposition here that I think ultimately the CDI team needs to weigh. In this case, I'm here to point out the risk. I trust the CDI maintainers to make the final call.

Reply:
After thinking about this a bit, I like the TTL semantics, but I think I would change the `deleteAfterCompletion` annotation to deleteAfterCompletion, and the value is a timestamp. The timestamp is calculated based on the time when the DV completed + TTL. This would allow for:
1. If TTL is not defined, don't add annotation
2. If TTL is 0 then the DV is immediately marked for deletion
3. If TTL > 0 then we can wait until the current time > timestamp we set.
Now existing DVs will simply fall in category 1 since there is no annotation, and the DVs will act as if TTL was not defined. The only problem I see with this is how do we reliable trigger the reconcile loop that will ultimately GC the DV if the current time > timestamp if the annotation exists. If for whatever reason the CDI control plane is down while the time expires, we would have to do some cleanup when it comes back up.

Reply:
@awels from your comment it seems an existing DV which is not completed (e.g. PVC was deleted) will be deleted or annotated for deletion once it is completed, which is not the expected behavior. Moreover, we wanted to allow user to explicitly add this annotation for older DVs which is also a bit non-trivial with your suggestion.

@davidvossel @mhenriks regarding the risky default `deleteAfterCompletion` DV annotation (if `dataVolumeTTL` is set), let's go with it and take a final decision later in the epic development.

Reply:
@awels when the DV controller reconciles a newly created DV and the TTL is set in CDIConfig, it sets the DV `deleteAfterCompletion`. Any problem with this behavior?

Reply:
Instead of adding a timestamp, @mhenriks suggested using DV mutating webhook for setting the `deleteAfterCompletion` if CDIConfig TTL is set.

Reply:
Since the mutating webhook will only act on DVs, and DVs won't work if the CDI API is down, that might be a simpler option than calculating timestamps.